File: leetcode/0376-wiggle-subsequence/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if (nums.empty()) return 0;

        int length = 1;
        int prevDiff = 0;
        for (int i = 1; i < (int)nums.size(); i++) {
            int diff = nums[i] - nums[i - 1];
            if (diff > 0 && prevDiff <= 0) {
                length++;
                prevDiff = 1;
            } else if (diff < 0 && prevDiff >= 0) {
                length++;
                prevDiff = -1;
            }
        }
        return length;
    }
};

File: leetcode/0376-wiggle-subsequence/solution.py
from typing import List


class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        if not nums:
            return 0

        length = 1
        prev_diff = 0
        for i in range(1, len(nums)):
            diff = nums[i] - nums[i - 1]
            if diff > 0 and prev_diff <= 0:
                length += 1
                prev_diff = 1
            elif diff < 0 and prev_diff >= 0:
                length += 1
                prev_diff = -1
        return length
